tfx/extensions/google_cloud_kubernetes/runner.py

```
@@ -0,0 +1,245 @@
+                                       unique_id=unique_id)
+  pod_names = _build_pod_names(num_workers=num_workers, unique_id=unique_id)
+  worker_hosts = ['{}:5000'.format(svc_name) for svc_name in service_names]
+
+  # TODO(ericlege): consider using a jinja2 template instead
+  for i in range(num_workers):
+    tf_config = json.dumps({
+        'cluster': {
+            'worker': worker_hosts
+        },
+        'task': {'type': 'worker', 'index': i}
+    })
+    pod = client.V1Pod(
+        metadata=client.V1ObjectMeta(
+            name=pod_names[i],
+            labels={
+                'name': 'training',
+                'id': unique_id,
+                'task': str(i),
+            },
+        ),
+        spec=client.V1PodSpec(
+            containers=[
+                client.V1Container(
+                    name='worker-pod',
+                    image=tfx_image,
+                    command=_COMMAND,
+                    args=job_args,
+                    security_context=client.V1SecurityContext(
+                        privileged=True,
+                    ),
+                    env=[
+                        client.V1EnvVar(
+                            name='TF_CONFIG',
+                            value=tf_config,
+                        ),
+                    ],
+                    ports=[
+                        client.V1ContainerPort(
+                            container_port=5000,
+                        ),
+                    ],
+                    resources=client.V1ResourceRequirements(
+                        limits={
+                            'nvidia.com/gpu': num_gpus_per_worker,
+                        },
+                    ) if num_gpus_per_worker > 0 else None,
+                ),
+            ],
+            restart_policy=kube_utils.RestartPolicy.NEVER.value,
+        ),
+    )
+    try:
+      api_instance.create_namespaced_pod(namespace='default', body=pod)
+    except ApiException as e:
+      logging.error(
```

Return after logging? Should this be a fatal instead / should we reraise the exception?

---

Agreed, changed to re-raise using the from keyword.